### PR TITLE
Enum choice using a GUI

### PR DESCRIPTION
**Pull request by sam543381**
This commits allows API's users to request a player to make a choice between multiple enum values by simply providing a callback and a function for translating each enum value to ItemStack shown in the GUI.

### CODE REVIEW

**Comment by AmauryCarrade** [comment]
Hi, I'm sorry but we cannot accept this PR as-is. Here's why.

First of all, we must support Java 7. If you look at the global Bukkit statistics regarding to the Java version, you'll see that more than 16% of the servers are still using Java 7 ([source](http://mcstats.org/global/), [archive](https://raw.carrade.eu/s/1478989094.png), so dropping the support is currently not an option. Plus, it would break some plugins using it, and the API wouldn't be very consistent with only one part of the API using lambdas like this.

On another subject, this could be generalized: instead of accepting an `Enum`, you should accept a `T[]`. With that, you can still pass `EnumClass.values()` or `EnumClass.class.getEnumConstants()` to use an enumeration.

To implement this, you really should take a look at the [`PaginatedTextView`](https://github.com/zDevelopers/zLib/blob/master/src/main/java/fr/zcraft/zlib/tools/commands/PaginatedTextView.java) ([concrete usage here](https://github.com/AmauryCarrade/BallsOfSteel/blob/d351d3af6eb5bf430228349f958ce0c4b88cc999/src/main/java/eu/carrade/amaury/BallsOfSteel/commands/SpheresCommand.java#L105)) regarding on the way of implementing things in a coherent manner (at least coherent with zLib), as it's pretty the same king of “config” needed. Here in zLib, we use callbacks only for zero-config tools (like `PromptGui`), and subclasses for other things (e.g. the scoreboard, the paginated text view, etc.).

---

**TL;DR**: use `T[]` instead of `Enum`, replace `Function` with subclass (you can still create the subclass inline), and make the whole thing compatible with Java 7.

Anyway thanks for your contribution, it's nice to see you want to contribute! I hope these few requests will not discourage you.

**Comment by sam543381** [comment]
Hi,
Thank you for your answer and especially the time you've spent writing explanations. I am going to release a new PR later fixing problems you've listed above.

Best,
Samuel.

**Comment by AmauryCarrade** [comment]
You don't have to send a new PR, just push some new commits to the sam543381/zLib branch :) .

**Comment by sam543381** [comment]
Ok, thank for the information (my first maybe successful pr is coming ... :D).

Best,
Samuel.



On Mon, Nov 14, 2016 at 4:53 PM +0100, "Amaury Carrade" <notifications@github.com<mailto:notifications@github.com>> wrote:


You don't have to send a new PR, just push some new commits to the sam543381/zLib branch :) .

-

**Comment by sam543381** [comment]
The GUI was reviewed, I am waiting for your opinion.

Best,
Samuel.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	 * @param purpose
+	 *            The purpose (gui's title)
+	 * @param datas
+	 *            An array of datas to display
+	 * @param renderer
+	 *            Interface for building gui's ItemStack from array values
+	 * @param closeOnChoice
+	 *            Close the interface when the player has choosen if true
+	 */
+	public <A> ArrayPromptGui(Callback<T> callback, Player player, String purpose, T[] datas, ItemRenderer<T> renderer,
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: I'm not sure about this. You're still using callbacks in the constructor instead of a real subclass. For the callback, I agree (cc @prokopyl thought on that?), but as for the item renderer, grmpf.

I suggest you to make the `ArrayPromptGui` class abstract, and to add a new abstract method:
```java
abstract public ItemStack getViewItem(T data);
```
(renamed following the previously established convention in the [explorer GUI](https://github.com/zDevelopers/zLib/blob/master/src/main/java/fr/zcraft/zlib/components/gui/ExplorerGui.java#L459-L465)).

So, instead of using it like:
```java
new ArrayPromptGui((object) -> doSth(object), player, "title", data, new ItemRenderer<T>()
{
    @Override
    public ItemStack build(T data) { /* ... */ }
}, true);
```
you will instead use it like this:
```java
new ArrayPromptGui((object) -> doSth(object), player, "title", data, true)
{
    @Override
    public ItemStack getViewItem(T data) { /* ... */ }
};
```

It's not really different, and the advantage is you can create a concrete child class for repeated use instead of uglying storing the rendered in a variable or, even worse, duplicating some code. Also, child classes could override the constructor to set some defaults :) .

Exactly like in the [`PaginatedTextView`](https://github.com/zDevelopers/zLib/blob/master/src/main/java/fr/zcraft/zlib/tools/commands/PaginatedTextView.java#L200-L288) ;) .

Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: Another possibilty to change this: use a builder pattern, [like this](https://github.com/AmauryCarrade/BallsOfSteel/blob/d351d3af6eb5bf430228349f958ce0c4b88cc999/src/main/java/eu/carrade/amaury/BallsOfSteel/commands/SpheresCommand.java#L105-L108), to configure every non-required option (i.e. everything except player), applying sensible defaults to unset options. The display would be done using a final method in the chain, like in the given example. The constructor would be arguments-less (like 0).

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import fr.zcraft.zlib.components.gui.ExplorerGui;
+import fr.zcraft.zlib.components.gui.Gui;
+import fr.zcraft.zlib.tools.Callback;
+
+/* Built from recommendations in PR https://github.com/zDevelopers/zLib/pull/14 */
+public class ArrayPromptGui<T> extends ExplorerGui<T> {
+
+	private Callback<T> cb;
+	private String purpose;
+	private T[] datas;
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: The plural of `data` is `data`, without `s`. Same for all other `datas` variables & parameters.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+import fr.zcraft.zlib.components.gui.ExplorerGui;
+import fr.zcraft.zlib.components.gui.Gui;
+import fr.zcraft.zlib.tools.Callback;
+
+/* Built from recommendations in PR https://github.com/zDevelopers/zLib/pull/14 */
+public class ArrayPromptGui<T> extends ExplorerGui<T> {
+
+	private Callback<T> cb;
+	private String purpose;
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: This should be called `title` (clearer), I think. And all parameters and variables named the same way.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java (hunk 1 above)
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: After a bit of reflection with @prokopyl, the callback should also be added as an abstract method, so even in Java 7 (where you have to create a whole anonymous class for this), the code is clearer and more coherent.
Sad for the lambdas expressions. But better in others points of views.

Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: You can also add both a callback in the constructor and as a method to override, the method defaulting to executing the stored callback if it's not `null`. *The Return of the Lambda*.

**Comment by sam543381** [comment]
Hello,
I remade the class. Do I need to remake a PR ?

Best,
Samuel.

**Comment by AmauryCarrade** [comment]
As before, push to the same branch and it will appears here. Pull requests are branch-based.

**Comment by sam543381** [comment]
Hello,
There is'nt more branches on my fork. I pushed on master, but ... nothing here.

Best,
Samuel.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 4]
```diff
@@ -7,87 +7,105 @@
...
+	 */
+	public abstract ItemStack getViewItem(T data);
+
+	/**
+	 * Called when player made a choice if no callback was provided
+	 * 
+	 * @param data
+	 */
+	public void onChoice(T data) {
+		System.err.println("Damn ! I'm not properly implemented : override me or use a callback.");
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: Prefer throwing an `UnsupportedOperationException` (or maybe `org.apache.commons.lang.NotImplementedException`) for these kind of errors.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 5]
```diff
@@ -7,87 +7,105 @@
...
+	 * @return The ItemStack to display
+	 */
+	public abstract ItemStack getViewItem(T data);
+
+	/**
+	 * Called when player made a choice if no callback was provided
+	 * 
+	 * @param data
+	 */
+	public void onChoice(T data) {
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: I'm not comfortable with the method name. `onClick` is better, I think.

File: src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
+package fr.zcraft.zlib.components.gui;
```
Comment (AmauryCarrade) [inline src/main/java/fr/zcraft/zlib/components/gui/ArrayPromptGui.java]: Please add this comment at the top of the file, just before the `package` statement.

```java

/*
 * Copyright or © or Copr. ZLib contributors (2015 - 2016)
 *
 * This software is governed by the CeCILL-B license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL-B
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-B license and that you accept its terms.
 */
```

**Comment by AmauryCarrade** [comment]
> There is'nt more branches on my fork. I pushed on master, but ... nothing here.
Well, the commit is there. I added a few last notes.

**Comment by AmauryCarrade** [comment]
Are you still there?

**Comment by sam543381** [comment]
I am but I have enough time to make these changes.

Best,
SamSam

**Comment by AmauryCarrade** [comment]
You mean, you **don't** have enough time? If so, I can make these for you, if you accept them (especially on your code using this).

**Comment by sam543381** [comment]
Yep, I would be happy if you do it.
